Updates to code contribution guide (#4551)
magento/devdocs: guides/v2.2/contributor-guide/contributing.md

When the Community Engineering Team works on reviewing the suggested changes, we will add a label to the issue to indicate certain information, like the status or who is working the issue. See [Labels](#labels) to learn more.

{:.bs-callout .bs-callout-tip}
Please refer to [Magento Contributor Agreement](http://www.magento.com/legaldocuments/mca) for detailed information about the License Agreement. All contributors are required to submit a click-through form to agree to the terms.

## GitHub and Two-Factor Authentication {#two-factor}

When setting up access and tokens for Magento GitHub repositories, we recommend adding Two-Factor Authentication (2FA) to enhance security. 2FA adds an additional layer of security beyond just a username/password when you access GitHub, to protect your private and public repositories.

Magento **requires all Partners** who contribute code to enable 2FA on their accounts. You can use a mobile device or 2FA application for added protection.

> For details, see [Configuring Two-Factor Authentication via a mobile app guide](https://help.github.com/articles/configuring-two-factor-authentication-via-a-totp-mobile-app/) to add 2FA protection to your GitHub account.

Reviewer: ```suggestion
```

Contributions can take the form of new components or features, changes to existing features, tests, documentation (such as developer guides, user guides, examples, or specifications), bug fixes, optimizations, or just good suggestions. We recommend also creating back-ports and up-ports as needed with your pull requests if the issue occurs in higher or lower line releases.

The Community Engineering Team reviews all issues and contributions submitted by the community developers. During the review we might require clarifications from the contributor. If there is no response from the contributor in two weeks (14 days) time, the issue is closed.

When the Community Engineering Team works on reviewing the suggested changes, we will add a label to the issue to indicate certain information, like the status or who is working the issue. See [Labels](#labels) to learn more.

{:.bs-callout .bs-callout-tip}

> Please refer to [Magento Contributor Agreement](http://www.magento.com/legaldocuments/mca) for detailed information about the License Agreement. All contributors are required to submit a click-through form to agree to the terms.

Reviewer 2: ```suggestion
Refer to [Magento Contributor Agreement](http://www.magento.com/legaldocuments/mca) for detailed information about the License Agreement. All contributors are required to submit a click-through form to agree to the terms.
```

Reviewer 3: Yes!, no begging allowed.

## GitHub and Two-Factor Authentication {#two-factor}

> When setting up access and tokens for Magento GitHub repositories, we recommend adding Two-Factor Authentication (2FA) to enhance security. 2FA adds an additional layer of security beyond just a username/password when you access GitHub, to protect your private and public repositories.

Reviewer 2: ```suggestion
When setting up access and tokens for Magento GitHub repositories, we recommend adding Two-Factor Authentication (2FA) to enhance security. 2FA adds an additional layer of security, beyond just a username/password, when you access GitHub, to protect your private and public repositories.
```

Reviewer 4: A Community Engineer tells me that 2FA is required for contributors.

> Magento **requires all Partners** who contribute code to enable 2FA on their Github accounts. You can use a mobile device or 2FA application for added protection. See [Configuring two-factor authentication](https://help.github.com/en/articles/configuring-two-factor-authentication) in the Github help.

Reviewer 2: ```suggestion
Magento **requires all Partners** who contribute code to enable 2FA on their GitHub accounts. You can use a mobile device or 2FA application for added protection. See [Configuring two-factor authentication](https://help.github.com/en/articles/configuring-two-factor-authentication) in the GitHub help.
```

1. Pull requests (PRs) must be accompanied by a meaningful description of their purpose. Comprehensive descriptions increase the chances that a pull request is merged quickly and without additional clarification requests.
1. Commits must be accompanied by meaningful commit messages.
1. PRs that include bug fixes must be accompanied by a step-by-step description of how to reproduce the bug.
1. PRs that include new logic or new features must be submitted along with:
  * Unit/integration test coverage (we will be releasing more information about writing test coverage in the near future).
  * Follow the [Magento DevDocs contribution work flow and guidelines](https://github.com/magento/devdocs/blob/master/.github/CONTRIBUTING.md) to submit proposed documentation updates to the [Magento DevDocs Github repository](https://github.com/magento/devdocs).
1. For large features or changes, [open an issue](https://github.com/magento/magento2/issues) to discuss your proposal first.  Notifying us in advance can prevent duplicate or unnecessary effort, and also offers an opportunity to get additional background information and help from other contributors. 

> 1. To report a bug, [open an issue](https://github.com/magento/magento2/issues), and follow the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines).

Reviewer 2: ```suggestion
1. To report a bug, [open an issue](https://github.com/magento/magento2/issues) and follow the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines).
```

1. Verify that all automated tests on your pull request pass successfully–all builds on [Travis CI](https://travis-ci.org/magento/magento2) must be green.

> ## Forks and Pull Requests

Reviewer 2: ```suggestion
## Forks and pull requests
```

> For complete information to start contributing to Magento projects, see the [Beginner Guides](https://github.com/magento/magento2/wiki/Getting-Started) on the Magento 2 repository. These guides help you:

Reviewer 2: ```suggestion
For complete information about contributing to Magento projects, see the [Beginner Guides](https://github.com/magento/magento2/wiki/Getting-Started) on the Magento 2 repository. These guides help you:
```

### What are forward-ports and back-ports?

As a best practice, we recommend creating back-ports and forward-ports for your code contributions.

**Back-ports** (or backports) contribute your code and fixes to previous release lines. Create a back-port if you fix an issue in the current release line that also affects an earlier release. For example, if you fix an issue in 2.3, create a pull request against the 2.2 
branch to apply your changes to that release. 

**Forward-ports** (or up-ports) contribute your code and fixes to a more current release line. Create a forward port if you fix an issue in an earlier release line that also affects a more current release. For example, if you have fixed an issue in 2.2, create a pull request against the latest release branch to address that issue in an upcoming minor release. If you contribute a fix to 2.2, create a pull request against the 2.3 release branch to apply your changes to that release. We recommend that contributors create a pull request to forward port every fix delivered to an older release line.

> ### Magento porting tool {#porting-tool}

Reviewer: ```suggestion
### Magento Porting Tool {#porting-tool}
```

## Porting code contributions across Magento versions {#porting}

> In order to keep consistency between Magento release lines (2.2, 2.3, etc), there are two techniques of code delivery: back-port and/or up-port. Every Magento Contributor, who wants to deliver their solution across all Magento versions, faces the same problem. How do you port fixes easily?

Reviewer 2: ```suggestion
To maintain consistency between Magento release lines (2.2, 2.3, etc), there are two techniques of code delivery: back-port and/or up-port. Every Magento contributor who wants to deliver their solution across all Magento versions faces the same problem. How do you port fixes easily?
```

### What are forward-ports and back-ports?

As a best practice, we recommend creating back-ports and up-ports for your code contributions.

**Back-ports** (or backports) contribute your code and fixes to a lower release line. You want to create a back-port if the issue exists in a lower release line. If you fixed an issue in a release line and there is a supported lower version, create a pull request to that lower release line to address the issue. For example, you may have contributed a fix to 2.3 and back-port to 2.2 and 2.1.

**Forward-ports** (or up-ports) contribute your code and fixes to a more current release line. Create a forward port if you fix an issue in an earlier release that also affects a more current release line. For example, if you have fixed an issue in 2.2, create a pull request to the latest release branch to address that issue in an upcoming minor release. If you contribute a fix to 2.2, create a pull request to forward port the change to the 2.3 branch. We recommend that contributors create a pull request to forward port every fix delivered to an earlier release line.

> ### Magento Porting Tool {#porting-tool}

Reviewer: ```suggestion
### Magento porting tool {#porting-tool}
```

> ### Magento Porting Tool {#porting-tool}

Reviewer: ```suggestion
### Magento porting tool {#porting-tool}
```

> **Back-ports** (or Backports) contribute your code and fixes to a lower release line. You want to create a back-port if the issue exists in a lower release line. If you fixed an issue in a release line and there is a supported lower version, create a pull request to that lower release line to address the issue. For example, you may have contributed a fix to 2.3 and back-port to 2.2 and 2.1.

Reviewer 2: ```suggestion
**Back-ports** (or Backports) contribute your code and fixes to a lower release line. Create a back-port if the issue exists in a lower release line. If you fixed an issue in a release line and there is a supported lower version, create a pull request to that lower release line to address the issue. For example, you may have contributed a fix to 2.3 and back-ports to 2.2 and 2.1.
```

> **Up-ports** (or Forwardports) contribute your code and fixes to a higher release line. You want to create an up-port if the issue exists in a higher release line. If you have an issue fixed in a non-latest release line, create a pull request to the latest branch in order to address that issue in an upcoming minor release. For example, you may have contributed a fix to 2.2 and up-port to 2.3. We recommend contributors create an up-port for every pull request delivered to a lower release line.

Reviewer 2: ```suggestion
**Up-ports** (or Forwardports) contribute your code and fixes to a higher release line. Create an up-port if the issue exists in a higher release line. If you have an issue fixed in a non-latest release line, create a pull request to the latest branch in order to address that issue in an upcoming minor release. For example, you may have contributed a fix to 2.2 and an up-port to 2.3. We recommend contributors create an up-port for every pull request delivered to a lower release line.
```

Reviewer: Same suggestion here -- "newer release" instead of "higher release" .  

Also, "if the issue exists" ==>  "If the issue affects a newer release line"

If you have an issue fixed in a non-latest release line ==> If you have fixed an issue in an earlier release line, create a pull request to the latest release branch to port the change to the upcoming minor release.

For example, when you create a fix to 2.2, create a pull request to up-port the fix to the 2.3 release branch.  We recommend that contributors create an up-port for every pull request delivered to an older release line.

> ### Magento Porting Tool {#porting-tool}

Reviewer 2: Case is not consistent. Ensure all are using the same style. I suggest sentence case.

> 1. Read the [issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) to learn how to create an issue that can be processed in a timely manner.

Reviewer 2: ```suggestion
1. Read the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) to learn how to create an issue that can be processed in a timely manner.
```

## Help triage issues  [![](https://www.codetriage.com/magento/magento2/badges/users.svg)](https://www.codetriage.com/magento/magento2) {#triage}

> In addition to contributing code, you can help to triage issues. This can include reproducing bug reports or asking for vital information, such as affected versions or instructions to reproduce bugs.  If you want to triage issues, you can begin by subscribing to [Magento on CodeTriage](https://www.codetriage.com/magento/magento2).

Reviewer 2: ```suggestion
In addition to contributing code, you can help triage issues. This can include reproducing bug reports or asking for vital information, such as affected versions or instructions to reproduce bugs.  If you want to triage issues, you can begin by subscribing to [Magento on CodeTriage](https://www.codetriage.com/magento/magento2).
```

## Labels applied by the Community Engineering team {#labels}

> We apply labels to public Pull Requests and Issues to help other participants retrieve additional information about current progress, component assignments, Magento release lines, and much more. The following information details global labels used in Magento 2 repositories and across Community Engineering contributions.

Reviewer 2: ```suggestion
We apply labels to public pull requests and issues to help other participants retrieve additional information about current progress, component assignments, Magento release lines, and much more. The following information details global labels used in Magento 2 repositories and across Community Engineering contributions.
```
